### DISCUSSION TITLE
Multiple Template Inheritance, cannot inherit a block across multiple templates

### DISCUSSION THREAD

**Discussion by OleksandrMyronchuk** [pallets/jinja#1944] [Q&A] [answered]
I'm working on a Flask application with Jinja templating. I have a base template (base.html) that defines several blocks for content and includes other templates like header.html, menu.html, and footer.html.

Within MyPage.html, I'm extending the base.html and defining a block named test1 with some content (test1). However, the rendered output only shows "test1=" with no content.

Here are the relevant code snippets:

App.py
```
...
@globals.app.route('/MyPage')
def MyPage():
    if os.path.exists(os.path.join(globals.app.root_path, 'templates/pages/MyPage.html')):
        return render_template('pages/MyPage.html')
    else:
        return jsonify({'error': 'Real time test trade page not found'}), 404
...
        
```
MyPage.html
```
{% extends 'base.html' %}

{% block test1 %}test1{% endblock %}
...
```
base.html
```
<!DOCTYPE html>
<html>
    {% include 'header.html' %}
    <body>
        <div class="container">
            <h1 class="text-center my-4">{% block header %}{% endblock %}</h1>
            {% include 'menu.html' %}
            {% block content %}{% endblock %}
        </div>
    </body>
    {% include 'footer.html' %}
</html>
```
header.html
```
<head>
    ...
    <title>{% block title %}{% endblock %}</title>
    ...
</head>
test1={% block test1 %}{% endblock %}
```
hierarchy
```
...
|
+---templates
|   |   body.html
|   |   footer.html
|   |   header.html
|   |   menu.html
|   |
|   \---pages
|           MyPage.html
...
```

I expected to see `test1=test1`, but now it is `test1=` (empty).

If I try to move `{% block test1 %}test1{% endblock %}` from MyPage.html to base.html, it works (`test1=test1`). However, in this case, the entire purpose of template programming is lost (I mean that you are forced to duplicate the code), so I would still prefer to use a template.

Could you please tell me how to fix my code so that the block `{% block test1 %}{% endblock %}` in header.html is rendered from the value set in the block `{% block test1 %}test1{% endblock %}` in MyPage.html?

**Comment by CheeseCake87** [comment, marked as answer]
I've created an example app based on what you've described, and I'm seeing the same result as you. I've looked over the docs and I can't find anything explaining the use of blocks with includes, but it's evident that blocks don't work with template includes.

Template includes will work with template variables though.

```html
{% for box in boxes %}
    {% include "render_box.html" %}
{% endfor %}
```

`{{ box }}` will be able to be seen by  `render_box.html`

A [Macro](https://jinja.palletsprojects.com/en/3.1.x/templates/#macros) or a [Context Processor](https://flask.palletsprojects.com/en/2.3.x/templating/#context-processors) is probably what you want to look into to solve the issue you're having.

For example, you'd replace `{% include 'header.html' %}` with a block:

`base.html`
```html
<!DOCTYPE html>
<html>
   {% block header %}{% endblock %}
    <body>
...
```

Then use a macro in `MyPage.html`

`header.html`
```html
{% macro header(value) -%}
    test1={{ value }}
{%- endmacro %}
```

`MyPage.html`
```html
{% import 'header.html' as header %}

...

{% block header %}
{{ header.header("hello") }}
{% endblock %}
```

Note: You may need to test the code above, I wrote this on the train quickly and haven't tested it xD
